library/std/src/thread/mod.rs

```diff
@@ -1257,112 +1252,113 @@ impl ThreadId {
 
 // This module ensures private fields are kept private, which is necessary to enforce the safety requirements.
 mod thread_name_string {
-    use core::str;
-
     use crate::ffi::{CStr, CString};
+    use crate::str;
 
     /// Like a `String` it's guaranteed UTF-8 and like a `CString` it's null terminated.
     pub(crate) struct ThreadNameString {
         inner: CString,
```

Comment:
If this was a `Cow<'static, CString>`, could the main thread reference a `"main\0"` string literal and thus avoid a few `else if main_thread::get() == Some(self.inner.id)`?

Reply:
The thread ID method has the advantage of working inside TLS destructors, so e.g. panicking in a TLS destructor on main now prints the correct name. And since we need the infrastructure anyway to detect the main thread in the first place, I figured this was better.

Reply:
> The thread ID method has the advantage of working inside TLS destructors, so e.g. panicking in a TLS destructor on main now prints the correct name. And since we need the infrastructure anyway to detect the main thread in the first place, I figured this was better.

Wait no, that's not related. I'll definitely undo the change from `ThreadName` then next time I update this PR.

Reply:
> The thread ID method has the advantage of working inside TLS destructors, so e.g. panicking in a TLS destructor on main now prints the correct name. And since we need the infrastructure anyway to detect the main thread in the first place, I figured this was better.

Sorry, that scenario wasn't the issue. The actual reason why this mustn't change is that there is now nothing stopping `thread::current` from being called before main, which would lead to an incorrect name being stored. The ID-based detection doesn't have this issue, because the name will be correct as soon as the runtime has initialized the ID.

Reply:
Ah, that is subtle. More comments would be good. :)